### ISSUE TITLE
(csharpsquid:S1244) Comparison of floating point values with 0 / zero should be allowed

### ISSUE THREAD

**Issue by tomsoftware** [SonarSource/sonar-dotnet#2326]
### Description

The rule "csharpsquid:S1244" should not complain on comparison of floating point values to exact zero.
https://rules.sonarsource.com/csharp/RSPEC-1244

The values: 0, NaN and Infinity are well defined state of a float point number so there is no need for checking this values within a delta. E.g the Sign(0) function is 0 and only 0 for 0!

### Repro steps

    float f = 0.0;
    if (f == 0) { // <-- S1244
    }

    if (f != 0) { // <-- S1244
    }

### Expected behavior

No sharpsquid:S1244 for zero checks. 

### Actual behavior

"Do not check floating point inequality with exact values, use a range instead."

### Known workarounds

    float f = 0.0;
    if ((System.Math.Sign(f) == 0)) {
    }

### Related information

* SonarC# SonarQube Version Version 6.7.6 (build 38781) (Web UI)

**Comment by Evangelink** [comment]
Hi @tomsoftware,

I have added your code sample on the latest version of SonarC# and we don't raise any issue. Could you confirm that the snippet is correct?

Cheers,
Amaury

**Comment by tomsoftware** [comment]
Hi @Evangelink ,

thanks for your response but I have to confirm this behavior

![grafik](https://user-images.githubusercontent.com/3290888/54144210-88bed100-442b-11e9-950f-1509d94bfb62.png)

so what can be the cause of this? do you have any idea?

**Comment by Evangelink** [comment]
You probably have a "too" old version of SonarC#/

**Comment by tomsoftware** [comment]
OK, I check the version (ConarC# : 7.11 (build 8083)) and I think it is the current available:

![grafik](https://user-images.githubusercontent.com/3290888/54280977-13710e80-4599-11e9-8d16-d988a5ab1445.png)


Also I'm not sure about the Test you added: I don't have a problem with e.g.
```
if (Math.Sign(f) == 0) {
}
```
but with

```
if (f == 0) {
}
```

So, the `Math.Sign(f)` only should demonstrate the special case a zero has in the a float point context  and the only way to avoid **S1244** I know.

`Math.Sign(f)  == 0`

equals

```
if ((f >= -0) && (f =< +0)) { // does not work by the way
}
```

My problem is: I don't know what range should I use to compare a floating point value with 0 to avoid **S1244** because there are almost "infinite" values around the 0 a float point number can represent. So I ask if you could add a case to allow float point comparing with 0 and don't rate them as a "bug".

**Comment by Evangelink** [comment]
I am not sure that I understood properly the point you are trying to make so please if my comment is not answering your question, feel free to let me know.

The usual way to solve the issue is to do:

```csharp
if (Math.Abs(f) < float.Epsilon) // use Epsilon from the type of 'f'
{
	// do something
}
```

or in case of a comparison:

```csharp
if (Math.Abs(x - y) < double.Epsilon)
{
    // do something
}
```

This `Epsilon` property is the smallest value that would make a difference between 2 numbers of a type. See [MSDN](https://docs.microsoft.com/en-us/dotnet/api/system.single.epsilon?view=netframework-4.7.2)

**Comment by tomsoftware** [comment]
I think it's going the wrong way. In my opinion the 0 is exactly defined and therefore comparison with exactly 0 should be allowed. But perhaps I am wrong.

I know enough workarounds to bypass the rule but such a workaround doesn't make the code easier to read / understand and this is actually the goal of Sonar.

By the way, Microsoft also does it:
https://github.com/dotnet/corefx/blob/5e3fe25d1236cd0192dd966f7286fe008d5cf875/src/Common/src/CoreLib/System/Convert.cs#L478

Greetings Thomas

**Comment by andrei-epure-sonarsource** [comment]
hi @tomsoftware . You can read this [SO answer](https://stackoverflow.com/a/485741/2261315) which mentions the fact that floating point numbers have precision error.

The only case where a comparison is safe is when you initialize yourself the number with `0` (or a different value) and then compare it with `0`, but that doesn't make a lot of sense. A variable can be the result of a floating point operation, which is likely to have a precision error.

For `Nan` and `Infinity` you can use [IsNaN ](https://docs.microsoft.com/en-us/dotnet/api/system.double.isnan?view=netframework-4.8), [IsPositiveInfinity](https://docs.microsoft.com/en-us/dotnet/api/system.double.ispositiveinfinity?view=netframework-4.8), etc. Unfortunately, the framework doesn't provide an `IsZero` method.

**Comment by tboby** [comment]
@andrei-epure-sonarsource Is there any chance of an escape hatch being added here? We very frequently *are* checking for an exactly assigned default value of `0`.

If you'd like an example of a competitor/alternative: JetBrains Rider only warns on equality when non-zero.

**Comment by tkouba** [comment]
Still not fixed. SonarQube 8.25.0.13952
